### PR TITLE
Restoring typing guards in the SMT encoding of recursive functions

### PR DESCRIPTION
**Pull request by nikswamy**
The story of this pull request is quite long. 

A couple of weeks ago, @joonwonc reported a fairly compact proof of false, later minimized further by @aseemr  to

```
module Test

let rec log (n:nat{n > 0}) :nat = if n = 0 then 1 + log n else 1

let bar (n:nat) = True

val foo :nat -> Type0
let foo j = j == 0 \/ bar (log j)

let test (j:nat) =
  assert (foo j);
  assert (j > 0) //effectively false from here on ...
```

The unsoundness dates back to some work I did on optimizing the SMT encoding 2.5 years ago: 26c05854262cfd2d9b37c546ac802b8e0fe7b20a

What I did then was to omit typing guards when encoding to SMT the definitions of F* functions (retaining them only for typing axioms, obviously). That means, a function like `log` above is encoded roughly (ignoring fuel instrumentation of recursive functions) as:

```
(declare-fun log (Term) Term)
(assert 
(forall ((x Term) 
    (! (= (log x)
            (ite (= x (BoxInt 0))
                  (op_Addition (BoxInt 1) (log x))
                  (Box 1)))
      :pattern ((log x)))  
```

The reasoning then was that this equation is guarded by a pattern and all occurrences of `log x` elsewhere in the SMT encoding will only be for non-zero instances of `x` (by F* typing), so, although this definition is ill-founded (esp. at 0), the SMT solver should never be able to prove that.

That turns out to not be true. Ultimately, relying on quantifier instantiation heuristics for soundness of the logical encoding is unwise.

Over the last couple of weeks @aseemr and I have been trying many ways to fix this problem without completely undoing that optimization. (Along the way we discovered a somewhat narrow unsoundness in Z3 too, which was reported and fixed by the Z3 authors immediately)

After several false starts, @Chris-Hawblitzel suggested to only retain the typing guards on recursive definitions, since non-recursive definitions cannot suffer from these kinds for ill-foundedness problems.

So, this pull request basically implements Chris' suggestion, retaining typing and precondition guards for recursive functions only. Along the way, there were other related bugs that had to be fixed, notably:

-- incorrect typing axioms for fuel-instrumented recursive functions
-- revising the encoding of reified functions so that we can encode their pre/post-conditions to SMT properly

This may look like a big diff, but many are to generated files. I would suggest looking only at

```
 git diff origin/master --name-only | grep -v hints | grep -v ocaml-output | xargs git diff origin/master
```

Some proofs needed to be changed in non-trivial ways, but thankfully not too many or too deeply. Mainly, some proofs involving recursive functions now need extra lemmas to prove the new typing guards, or rely on normalization instead of SMT to unroll recursive definitions.

### CODE REVIEW

**Comment by aseemr** [comment]
Here are the links for source files:

https://github.com/FStarLang/FStar/pull/1601/files#diff-5fd118520e4f76d7b3e4d20518f010de

https://github.com/FStarLang/FStar/pull/1601/files#diff-5c1447402ef473f97167535423398c76

https://github.com/FStarLang/FStar/pull/1601/files#diff-27285a959f05d6176d5a64e2da907bd0

File: src/smtencoding/FStar.SMTEncoding.Encode.fs [hunk 1]
```diff
@@ -528,73 +502,78 @@ let encode_top_level_let :
...
+      let binders, body, comp = aux t e in
+      let binders, body, comp =
+          if is_reifiable_comp tcenv comp
+          then let comp = reify_comp tcenv comp U_unknown in
+               let body = TcUtil.reify_body tcenv body in
+               let more_binders, body, comp = aux comp body in
+               binders@more_binders, body, comp
+          else binders, body, comp
+      in
+      binders, U.ascribe body (BU.Inl (U.comp_result comp), None), comp
```
Comment (aseemr) [inline src/smtencoding/FStar.SMTEncoding.Encode.fs]: What's up with the `U.ascribe` here? Won't it be dropped anyway going forward?

**Comment by aseemr** [comment]
Left a minor comment. Otherwise looks good to me. Thanks.

File: src/smtencoding/FStar.SMTEncoding.Encode.fs [hunk 2]
```diff
@@ -502,76 +502,67 @@ let encode_top_level_let :
...
+//       * bool         //if set, we should generate a curried application of f
     =
       (* The input type [t_norm] might contain reifiable computation type which must be reified at this point *)
-      let get_result_comp c =
-          if is_reifiable_comp env.tcenv c
-          then S.mk_Total <| reify_comp ({env.tcenv with lax = true}) c U_unknown
-          else c
-      in
+
+      let tcenv = {env.tcenv with lax=true} in
```
Comment (mtzguido) [inline src/smtencoding/FStar.SMTEncoding.Encode.fs]: I don't think `lax` makes a difference here. Can we just define `tcenv = env.tcenv`?

**Comment by mtzguido** [comment]
Looks good to me too, with a tiny nit above

**Pull request merged**